rfc/rfc-37/rfc-37.md

```diff
@@ -0,0 +1,168 @@
+due to loading of bloom from large number of data files and due to throttling issues with some of the cloud stores. We are proposing to 
+build a new Metadata index (metadata table based bloom index) to boost the performance of existing bloom index. 
+
+## Background
+HoodieBloomIndex is used to find the location of incoming records during every write. This will assist Hudi in deterministically 
+routing records to a given file group and to distinguish inserts vs updates. This bloom index relies on (min, max) values 
+of records keys and bloom indexes in base file footers to find the actual record location. In this RFC, we plan to 
+build a new index on top of metadata table which to assist in bloom index based tagging. 
+
+## Design
+HoodieBloomIndex involves the following steps to find the right location of incoming records
+1. Load all interested partitions and fetch data files. 
+2. Find and filter files to keys mapping based on min max in data file footers.
+3. Filter files to keys mapping based on bloom index in data file footers. 
+4. Look up actual data files to find the right location of every incoming record.
+
+As we could see from step 1 and 2, we are in need of min and max values for "_hoodie_record_key" and bloom filter for 
+all data files to perform the tagging. In this design, we will add these to metadata table and the index lookup 
+will look into these metadata table partitions to deduce the file to keys mapping. 
+
+To realize this, we are adding two new partitions namely, `column_stats` and `bloom_filter` to metadata table.  
+
+Why metadata table: 
+Metadata table uses HFile to store and retrieve data. HFile is an indexed file format and supports random lookups based on 
+keys. Since, we will be storing stats/bloom for every file and the index will do lookups based on files, we should be able to 
+benefit from the faster lookups in HFile. 
+
+<img src="metadata_index_1.png" alt="High Level Metadata Index Design" width="800"/>
+
+Following sections will talk about different partitions, key formats and then dive into the data and control flows.
+
+### Column_Stats partition:
+"Column_stats" will be discussed in depth in RFC-27, but in the interest of this RFC, Column_stats partition stores 
+statistics(min and max value) for `__hoodie_record_key` column for all files in the Hudi data table. 
+
+High level requirement for this column_stats partition are:
+Given a list of record keys, partition paths and file names, find the possibly matching file names based on
+`__hoodie_record_key` column stats. 
+
+To cater to this requirement, we need to ensure our keys in Hfile are such that we can do pointed lookups for a given data file.
+Below picture gives a pictorial representation of Column stats partition in metadata table. 
+
+<img src="metadata_index_col_stats.png" alt="Column Stats Partition" width="600"/>
+
+We have to encode column names, filenames etc to IDs to save storage and to exploit compression. We will update the RFC 
```

Comment:
encoding column names, partition names may not be required as HFile compresses blocks of key-value data. So repeated string of column names, etc will compress well.